File: app/models/alert.rb
# frozen_string_literal: true

#
# Copyright (C) 2011 - present Instructure, Inc.
#
# This file is part of Canvas.
#
# Canvas is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, version 3 of the License.
#
# Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License along
# with this program. If not, see <http://www.gnu.org/licenses/>.
#

class Alert < ActiveRecord::Base
  belongs_to :context, polymorphic: [:account, :course]
  has_many :criteria, class_name: "AlertCriterion", dependent: :destroy, autosave: true

  serialize :recipients

  validates :context_id, presence: true
  validates :context_type, presence: true
  validates :criteria, presence: true
  validates_associated :criteria
  validates :recipients, presence: true

  before_save :infer_defaults

  def find_role_by_name(role_name)
    context.is_a?(Account) ? context.get_account_role_by_name(role_name) : context.account.get_account_role_by_name(role_name)
  end

  def resolve_recipients(student_id, teachers = nil)
    include_student = false
    include_teachers = false
    admin_role_ids = []
    recipients.try(:each) do |recipient|
      case recipient
      when :student
        include_student = true
      when :teachers
        include_teachers = true
      when String
        admin_role_ids << find_role_by_name(recipient).id
      when Hash
        admin_role_ids << recipient[:role_id]
      else
        raise "Unsupported recipient type!"
      end
    end

    recipients = []

    recipients << student_id if include_student
    recipients.concat(Array(teachers)) if teachers.present? && include_teachers
    if context_type == "Account" && !admin_role_ids.empty?
      recipients.concat context.account_users.active.where(role_id: admin_role_ids).distinct.pluck(:user_id)
    end
    recipients.uniq
  end

  def infer_defaults
    self.repetition = nil if repetition.blank?
  end

  def as_json(*)
    converted_recipients = recipients.to_a.map do |recipient|
      case recipient
      when String
        find_role_by_name(recipient).id
      when Hash
        recipient[:role_id]
      else
        ":#{recipient}"
      end
    end
    {
      id:,
      criteria: criteria.map { |c| c.as_json(include_root: false) },
      recipients: converted_recipients,
      repetition:
    }.with_indifferent_access
  end

  def criteria=(values)
    if values[0].is_a? Hash
      values = values.map do |params|
        if params[:id].present?
          id = params.delete(:id).to_i
          criterion = criteria.to_ary.find { |c| c.id == id }
          criterion.attributes = params
        else
          criterion = criteria.build(params)
        end
        criterion
      end
    end
    criteria.replace(values)
  end
end

File: client_apps/canvas_quizzes/apps/common/js/components/alert.js
/** @jsx React.DOM */
/*
 * Copyright (C) 2014 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

define(function(require) {
  var React = require('old_version_of_react_used_by_canvas_quizzes_client_apps')
  var classSet = require('../util/class_set')
  var $ = require('jquery')

  // TODO: use $.fn.is(':in_viewport') when it becomes available
  var isInViewport = function(el) {
    var $el = $(el)
    var $window = $(window)

    var vpTop = $window.scrollTop()
    var vpBottom = vpTop + $window.height()
    var elTop = $el.offset().top
    var elBottom = elTop + $el.height()

    return vpTop < elTop && vpBottom > elBottom
  }

  /**
   * @class Components.Alert
   *
   * A Bootstrap alert-like component.
   */
  var Alert = React.createClass({
    propTypes: {
      /**
       * @cfg {Boolean} [autoFocus=false]
       *
       * If true, the alert will auto-focus itself IF it is visible within
       * the viewport (e.g, the user can currently see it.).
       *
       * This is useful to force ScreenReaders to read the notification.
       */
      autoFocus: React.PropTypes.bool
    },

    getDefaultProps: function() {
      return {
        type: 'danger',
        autoFocus: false
      }
    },

    componentDidMount: function() {
      if (this.props.autoFocus) {
        var myself = this.getDOMNode()

        if (isInViewport(myself)) {
          setTimeout(function() {
            myself.focus()
          }, 1)
        }
      }
    },

    render: function() {
      var className = {}

      className['alert'] = true
      className['alert-' + this.props.type] = true

      return (
        <div
          tabIndex="-1"
          aria-role="alert"
          aria-live="assertive"
          aria-relevant="all"
          onClick={this.props.onClick}
          className={classSet(className)}
          children={this.props.children}
        />
      )
    }
  })

  return Alert
})
